initial draft of an introduction to MAF

ehneilsen opened the pull request:
I think I have taken care of your suggestions, and then done a little more work with style checking/linting using DMs preferred tools and settings (which they are now applying to notebooks as well).

rhiannonlynne commented:
In general, for tutorial notebooks I'd recommend explaining why you're using additional modules or tasks, etc. 
So the "import these to run the notebook cell" (cell 4) adds a lot of extra modules without clear uses -- if you're pretty familiar with python, it's obvious which ones of these are useful for what and when you *don't* need them, but otherwise, you'll find people adding these extra modules and steps when they don't need them.

In particular, I would be really clear about using tempfile and associated things -- most people using the notebook as a template will want to save their outputs permanently, not just temporarily, so it would be good to be really clear on why you're using that (vs. just saving to disk). 

Second, we provide the current baseline simulation in the rubin_sim_data_download -- if people are using rubin_sim, they will already have this file on disk. It would be better to demonstrate how to get and use that file, than add much overhead getting and downloading different files from ncsa to each notebook (especially when these cells tend to look a bit complicated, and add some extra overhead into the setting of the opsim names and file locations). So, in general, especially if it's possible to just demonstrate the notebook on a single database, I would suggest using those files instead and making the whole thing more targeted at what you're trying to highlight (plotting, running simple metrics, etc.).

ehneilsen commented:
How many different simulations are provided in rubin_sim_data_download? The tutorial includes a section on comparing two different ones, so I need to different sims to work with. (In other notebooks, I use more than two.)

It might be good to introduce the use of the baseline simulation they already have, but I also think in might be confusing to have the notebook obtain simulations in two different ways.

In any case, I'll go through and audit/explain the imports I do. Some may be left over from earlier drafts, and no longer even needed.

rhiannonlynne commented:
Cell 22-26 ... so, have you seen that you can set up a metric bundle and then tell that to read the file back in? 
This would avoid having to manipulate a numpy masked array directly, or access the header information in the npz file. 
It's better to encourage people to use this, I think, than get them to access the contents of the npz file directly.

Example:
```
my_bundle = maf.createEmptyMetricBundle()
my_bundle.read('my_filename')
```
my_bundle then has the slicer properties of the original slicer and the metric values of the original metric bundle. 
Depending on when it was saved and when the summary stats were calculated, it may or may not have summaryValues or plotDict information saved and restored appropriately as well.

I could've sworn we had a demo of this in one of the old tutorial notebooks, but now I can't find it .. so I guess it's not unexpected this is not super obvious.

rhiannonlynne commented:
You might consider removing some of the more complicated areas addressed in this intro notebook which are addressed in more depth in later notebooks (i.e. customizing plotting after the initial plot creation) and also linking to other notebooks which go into those areas in more depth (i.e. when introducing plotting, link to the plot notebook).

ehneilsen commented:
I discovered rereading bundles in the code after I write this notebook, while writing one of the others, and forgot I could go back and use it here.

rhiannonlynne commented:
The section on making maps says that they're basically the same but use different metrics .. okay, so this is our naming problem but it's pretty hard to change here. 
There is a bit of a maybe subtle difference in 'metric' (measuring something, how people use it) and "Metric" (our particular class). So, for example, you can actually use the same Metric to make a map as to evaluate a single value -- the MaxMetric will tell you the max seeing values over the sky or the max seeing value for the entire survey. In this case, the difference is the Slicer that is used. 
It's something to keep in mind when talking to people, because we've created this confusing situation (for historic reasons, it seemed like a good thing at the time)

> How many different simulations are provided in rubin_sim_data_download? The tutorial includes a section on comparing two different ones, so I need two different sims to work with. (In other notebooks, I use more than two.)
> 
> It might be good to introduce the use of the baseline simulation they already have, but I also think in might be confusing to have the notebook obtain simulations in two different ways.
> 
> In any case, I'll go through and audit/explain the imports I do. Some may be left over from earlier drafts, and no longer even needed.

You might also consider whether you *need* to compare two simulations in the introduction to MAF, or if running metrics on a single simulation would be sufficient when there is a "compare two simulations" and also "run metrics in bulk on multiple simulations" later. 

(there is only one, the current baseline, provided in the rubin_sim_data download)

rhiannonlynne commented:
The section on making maps says that they're basically the same but use different metrics .. okay, so this is our naming problem but it's pretty hard to change here. 
There is a bit of a maybe subtle difference in 'metric' (measuring something, how people use it) and "Metric" (our particular class). So, for example, you can actually use the same Metric to make a map as to evaluate a single value -- the MaxMetric will tell you the max seeing values over the sky or the max seeing value for the entire survey. In this case, the difference is the Slicer that is used. 
It's something to keep in mind when talking to people, because we've created this confusing situation (for historic reasons, it seemed like a good thing at the time and there was strong support from people doing the initial design reviews to set it up this way) and it's worthwhile to be extra specific when talking about metrics and Metrics. 
So, for example, this text:
```Example 3: Mapping the depth in r
Making maps follows a similar procedure, but uses different metrics.```
could be
```Example 3: Mapping the depth in r
Making maps follows a similar procedure, but uses different Slicers and potentially different Metrics.```

(I think basically what people usually think of as a 'metric' is equivalent to our "MetricBundle" -- a combo of a metric, slicer, and sql constraint).

rhiannonlynne commented:
One thing to consider that may be helpful (and we did not do this with the original tutorials) is to envision the tutorials as a series. 
Step 1: intro -- what is MAF and what are Metrics, Slicers, and how do they work with sqlconstraints to make a MetricBundle that then evaluates a "thing" we want to know about the simulation
Step 2: plotting
Step 3: running metrics on. more than one simulation (and possibly comparing outputs)
Step 4: writing a new metric

In this context, it may make even more sense to avoid temporary directories and reuse the previously calculated metric output for the later notebooks. (we could also provide a download link for pre-calculated metric npz files).

ehneilsen commented:
I was envisioning the tutorials very differently. 

I see two use cases for tutorial notebooks:
1. Self-teaching, in which someone finds them online and tries to follow them on their own.
2. Part of a "school" or tutorial session, usually a group with instructors that walk the students through the material.

Although similar in some ways, I think the two are quite different in others. In the 2nd case, most students will at least see most of the notebooks. But, I was mostly assuming the 1st use case in the notebooks I have been writing. In this case, I think most people will just follow the first introductory notebook, and then attempt to do whatever it is they want to use MAF for, only looking at any others when they run into something they can't figure out by trial and error or reading help pages.

So, I had two goals for the introductory notebook:
1. It should cover material that all (or at least most) users will need to see. Most users will need to make plots and compare simulations (otherwise, what good is calculating metrics?), so I covered that.
2. Provide enough conceptual background and mental "hooks" into the architecture that they can effectively make use of reference material (help strings, etc.). That is, they should know enough that they can learn more.

Most people who use MAF significantly will need some of the other tutorials, but not everyone will need the same ones, and often not right away. So, I have been assuming that by default each tutorial will be seen in isolation, and needs to be as "stand-alone" as possible.

ehneilsen commented:
I agree that we will have some users who only know basic python, but I think we can also expect that there will be quite advanced users as well. Python seems to have "taken over" astronomy in the past decade, and many users (particularly recent PhDs) will have pretty extensive experience with python and its modules (e.g. numpy, matplotlib, and pandas), but may be more accustomed to tools and idioms from outside of Rubin (astropy, pyraf, and modules independently developed for other projects), and get frustrated when these skills aren't easy to apply.

rhiannonlynne commented:
> I was envisioning the tutorials very differently.
> 
> I see two use cases for tutorial notebooks:
> 
> 1. Self-teaching, in which someone finds them online and tries to follow them on their own.
> 2. Part of a "school" or tutorial session, usually a group with instructors that walk the students through the material.
> 
> Although similar in some ways, I think the two are quite different in others. In the 2nd case, most students will at least see most of the notebooks. But, I was mostly assuming the 1st use case in the notebooks I have been writing. In this case, I think most people will just follow the first introductory notebook, and then attempt to do whatever it is they want to use MAF for, only looking at any others when they run into something they can't figure out by trial and error or reading help pages.
> 
> So, I had two goals for the introductory notebook:
> 
> 1. It should cover all material that all (or at least most) users will need to see. Most users will need to make plots and compare simulations (otherwise, what good is calculating metrics?), so I covered that.
> 2. Provide enough conceptual background and mental "hooks" into the architecture that they can effectively make use of reference material (help strings, etc.). That is, they should know enough that they can learn more.
> 
> Most people who use MAF significantly will need some of the other tutorials, but not everyone will need the same ones, and often not right away. So, I have been assuming that by default each tutorial will be seen in isolation, and needs to be as "stand-alone" as possible.

Yeah, the only problem with this is that 
(a) you tend to end up repeating contents so that each notebook can cover all of the relevant areas 
(b) things end up being hard to find, because of repeats and ending up including quite a bit of things in each one

At least, that is my thought from creating the series we had previously, where we tried the approach of making each one stand-alone and cover a particular topic. In the end, it seemed like people couldn't find things. If you create a series, at least you have a story that you're following and that others could follow. 

That said, I would talk to people in the CET or stack club or the Outreach/education group to see if they have further thoughts. In the end, I suspect either approach is all right but that indexing (and having a higher-level description of the contents of each notebook) and tracking gets much more important with a stand-alone series. We could try that again though.

rhiannonlynne commented:
> I agree that we will have some users who only know basic python, but I think we can also expect that there will be quite advanced users as well. Python seems to have "taken over" astronomy in the past decade, and many users (particularly recent PhDs) will have pretty extensive experience with python and its modules (e.g. numpy, matplotlib, and pandas), but may be more accustomed to tools and idioms from outside of Rubin (astropy, pyraf, and modules independently developed for other projects), and get frustrated when these skills aren't easy to apply.

Worth a try pitching the notebooks to a higher level, so sure.

ehneilsen commented:
I think indexing will be important either way. People won't want to reread the whole story each time they want to find something, and many people won't be willing to read the whole thing through even once.

Whatever the approach, I think there will need to be a "table of contents" notebook with links to all the others, with text describing what they cover and listing examples. Even when people do go through all the notebooks, as in a "school", they will need to be able to find stuff quickly for reference when they actually need to use it.

ehneilsen commented:
> Worth a try pitching the notebooks to a higher level, so sure.

I think your initial suggestion that I make them friendlier to those will less understanding was a good point, but I don't want to frustrate those who know more, either. How about I go back and see if I can't pitch each section to the minimum level that makes sense for that section?

rhiannonlynne commented:
Indexing/linking within the notebooks is good. Note that these links don't seem to work with notebooks viewed in GitHub, but do if you view them with nbviewer (see https://github.com/lsst-pst/survey_strategy/blob/main/fbs_2.0/Other%20Simulation%20Footprints.ipynb as an example). 

In case this wasn't what you were already planning, my suggestion would be to list and link the notebooks themselves in the README for the tutorial directory, rather than another notebook (I found the maintenance of a separate notebook was tedious (https://github.com/LSST-nonproject/sims_maf_contrib/blob/master/tutorials/Index.ipynb) and also harder to find, although it was nice when you were working with the whole repo on your own machine.  It's also easier to check that new tutorial notebooks are added to the list that way.

ehneilsen commented:
I think I have addressed the "lower level" comments now:

- Except for MAF itself, I moved the imports into the cells in which they are used. If this were a python source code file, it would be a stylistically poor choice (and it does make notebook development a bit more error-prone), but it should make clear to those with less python experience exactly what is going on in any given cell.
- I generally simplified the way I was handling run files and names, using single values variable instead of dictionaries. This makes the notebook less extendable and flexible, but it should make it easier to follow, particularly for those less familiar with python.
- I switched to using the baseline database from the rubin_sim simulation in all cases except for the one that does a comparison between to. In the latter, everything that involves downloading the other database is isolated into its own section right next to the only place where it is used, so hopefully this should minimize any confusion.
- I fixed the use of the word "metric" at the start of section 3.

I have not done anything yet to address the higher-level comments, concerning whether the comparison and plot customization belong in the introductory notebook at all, which relates to the overall strategy for organizing the notebooks and other documentation. I plan to follow your suggestion and find a CET slack to post a question on, probably pointing to this discussion.

In thinking about this, a third organization possibility has occurred to me: separate each and every example into its own notebook such that it is more a "collection of explained examples" than a tutorial per se, such that the current notebook becomes something like 5 or 6 separate notebooks plus an all-text "basic architecture" markdown document with links to the example notebooks.
